src/codegen/traverse.lisp

```diff
@@ -0,0 +1,359 @@
+  (declare (values action-list &optional))
+  (load-time-value
+   (list
+    (action (:traverse node-abstraction node traverse bound-variables)
+      (make-node-abstraction
+       :type (node-type node)
+       :vars (node-abstraction-vars node)
+       :subexpr (funcall traverse
+                         (node-abstraction-subexpr node)
+                         (append (node-abstraction-vars node) bound-variables))))
+    (action (:traverse node-let node traverse bound-variables)
+      (let ((new-bound-variables (append
+                                  (mapcar #'car (node-let-bindings node))
+                                  bound-variables)))
+        (make-node-let
+         :type (node-type node)
+         :bindings (loop :for (name . node) :in (node-let-bindings node)
+                         :collect (cons name (funcall traverse node new-bound-variables)))
+         :subexpr (funcall traverse (node-let-subexpr node) new-bound-variables))))
+    (action (:traverse node-match node traverse bound-variables)
+      (make-node-match
+       :type (node-type node)
+       :expr (funcall traverse (node-match-expr node) bound-variables)
+       :branches (mapcar
+                  (lambda (branch)
+                    (make-match-branch
+                     :pattern (match-branch-pattern branch)
+                     :body (funcall traverse
+                                    (match-branch-body branch)
+                                    (append (pattern-variables (match-branch-pattern branch))
+                                            bound-variables))))
+                  (node-match-branches node))))
+    (action (:traverse node-dynamic-extent node traverse bound-variables)
+      (let ((new-bound-variables (cons (node-dynamic-extent-name node) bound-variables)))
+        (make-node-dynamic-extent
+         :type (node-type node)
+         :name (node-dynamic-extent-name node)
+         :node (funcall traverse (node-dynamic-extent-node node) new-bound-variables)
+         :body (funcall traverse (node-dynamic-extent-body node) new-bound-variables))))
+    (action (:traverse node-bind node traverse bound-variables)
+      (let ((new-bound-variables (cons (node-bind-name node) bound-variables)))
+        (make-node-bind
+         :type (node-type node)
+         :name (node-bind-name node)
+         :expr (funcall traverse (node-bind-expr node) new-bound-variables)
+         :body (funcall traverse (node-bind-body node) new-bound-variables)))))
+   t))
+
+(defun traverse-with-binding-list (node actions)
+  "Traverse `node` while keeping track of a list of `bound-variables`
+which gets passed as an argument whenever a function from `actions`
+gets called."
+  (declare (type node        node)
+           (type action-list actions)
+           (values node &optional))
+  (traverse node
+            (append actions (make-binding-list-actions))
+            nil))
+
+;;;
+;;; Examples
+;;;
+
+(defun count-nodes (node)
```

Comment:
I pulled the general `traverse` stuff out to its own file and added these examples which I hope could be seen as `dead simple`

Reply:
skull emoji